src/ethereum/ethereum-reader.ts

```diff
@@ -194,18 +199,34 @@ export class EthereumReader {
         return this.web3.eth.getBlockNumber();
     }
 
-    async getRefTime(blockNumber: number | 'latest'): Promise<number | null> {
+    async getRefTime(blockNumber: number | 'latest'): Promise<number> {
         const block = await this.web3.eth.getBlock(blockNumber);
-        return block && toNumber(block.timestamp);
+        if (!block) {
+            throw new Error(`web3.eth.getBlock for ${blockNumber} return empty block`);
+        }
+        return toNumber(block.timestamp);
     }
 
+    // safe, if fails tries to decrease page size with multiple requests until it works
     async getPastEvents<T extends EventName>(
         eventName: T,
         { fromBlock, toBlock }: PastEventOptions
     ): Promise<Array<EventTypes[T]>> {
         const web3Contract = await this.connect(contractByEventName(eventName));
         return await getEventsPaged(web3Contract, eventName, fromBlock, toBlock, toBlock - fromBlock);
     }
+
+    // unsafe, throws error if fails, caller needs to decrease page size if needed
+    async getPastEventsUnsafe<T extends EventName>(
```

Comment:
this is a simplified form of `getPastEvents` for one block. consider accepting just `block` instead of a range, then you don't need the cryptic "unsafe" suffix. or, alternately, just use `getPastEvents` with it's overhead - it's negligible for the case where `fromBlock == toBlock`

Reply:
Good point, the name `getPastEventsUnsafe` is unclear. What I plan to do is retire the existing paged version of `getPastEvents` entirely. It has off-by-one bugs. I plan to move the retry auto page logic into the EventFetcher. To prepare for that, I renamed `getPastEventsUnsafe` to `getPastEvents` since it's a direct wrapper of the `web3` function under the same name. And I renamed `getPastEvents` to `getPastEventsAutoPaged` and added a TODO to retire it.